runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -0,0 +1,288 @@
+ * assumption that the address space of a stream can be traversed forward and backwards.
+ *
+ * The abstract method 'findAddresses' defines how to traverse the space of unexplored addresses,
+ * i.e, addresses between the latest synced point (oldTail) and the latest update (newTail).
+ * For example, the implementation of BackpointerAddressSpace assumes the use of backpointers
+ * to determine the previous log entry for a stream.
+ *
+ * Created by amartinezman on 4/23/18.
+ */
+public abstract class StreamAddressSpace implements IStreamAddressSpace {
+
+    protected CopyOnWriteArrayList<Long> addresses;
+    protected UUID streamId;
+    protected CorfuRuntime runtime;
+    protected int maxInd;
+    protected int currInd;
+    protected StreamOptions options;
+
+    public StreamAddressSpace(UUID id, CorfuRuntime runtime) {
+        this.streamId = id;
+        this.runtime = runtime;
+        this.addresses = new CopyOnWriteArrayList();
+        this.currInd = -1;
+        this.maxInd = -1;
+    }
+
+    @Override
+    public void setStreamOptions(StreamOptions options) {
+        this.options = options;
+    }
+
+    @Override
+    public void reset() {
+        currInd = -1;
+    }
+
+    @Override
+    public void seek(long address) {
+
+        int tmpPtr = addresses.indexOf(address);
+
+        if (tmpPtr != -1) {
+            currInd = tmpPtr;
+            return;
+        }
+
+        throw new RuntimeException("Couldn't seek " + streamId + " to address " + address);
+    }
+
+    @Override
+    public long getMax() {
+        if (maxInd == -1) {
+            return Address.NON_ADDRESS;
+        } else {
+            return addresses.get(maxInd);
+        }
+    }
+
+    @Override
+    public long getMin() {
+        if (addresses.size() == 0) {
+            return Address.NON_ADDRESS;
+        } else {
+            return addresses.get(0);
+        }
+    }
+
+    @Override
+    public long getLastAddressSynced() {
+        return getMax();
+    }
+
+    @Override
+    @Synchronized
+    public long next() {
+        if (currInd + 1 > maxInd) {
+            return Address.NON_ADDRESS;
+        } else {
+            currInd++;
+            return addresses.get(currInd);
+        }
+    }
+
+    @Override
+    @Synchronized
+    public long previous() {
+        if (currInd - 1 >= 0) {
+            currInd--;
+            return addresses.get(currInd);
+        } else {
+            currInd = -1;
+            return Address.NON_ADDRESS;
+        }    }
+
+    @Override
+    public List<Long> remainingUpTo(long limit) {
+        List<Long> res = new ArrayList<>(100);
+        while (hasNext() && addresses.get(currInd + 1) <= limit) {
+            res.add(addresses.get(currInd + 1));
+            currInd++;
+        }
+        return res;
+    }
+
+    @Override
+    public void addAddresses(List<Long> addresses) {
+        this.addresses.addAll(addresses);
+        Collections.sort(this.addresses);
```

Comment:
I believe this is unnecessary expensive, why do you need to sort every time you resolve a new delta? 

Also, can you add a sanity check to make sure the boundaries are correct (addresses[0] > this.addresses[last ind])
and maybe during tail traversal you can make sure that the backpointer is always decreasing.